perceval/client.py

```diff
@@ -0,0 +1,185 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
+#
+# Authors:
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+import time
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+from .errors import RateLimitError
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClient:
+    """Abstract class for HTTP clients.
+
+    Base class to query data sources.
+
+    Derivated classes have to implement `call`, `params`, `headers`
+    `pagination`, `has_retry` and `has_sleep` methods. Otherwise, `NotImplementedError`
+    exception will be raised.
+
+    To track which version of the client was used during the fetching
+    process, this class provides a `version` attribute that each client
+    may override.
+
+    :param has_rest: rest between calls
+    :param has_retry: retry calls in case of data source failures
+    :param has_pagination: client returns paginated items from the data source
+    """
+    version = '0.1'
+
+    MIN_RATE_LIMIT = 10
+    MAX_RATE_LIMIT = 500
+
+    DEFAULT_SLEEP_TIME = 1
+    MAX_RETRIES = 5
+
+    MAX_RETRIES_ON_CONNECT = 5
+    MAX_RETRIES_ON_READ = 5
+    MAX_RETRIES_ON_REDIRECT = 5
+    MAX_RETRIES_ON_STATUS = 5
+
+    RATE_LIMIT_HEADER = "X-RateLimit-Remaining"
+    RATE_LIMIT_RESET_HEADER = "X-RateLimit-Reset"
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [503, 413, 429]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, sleep_for_rate=False, min_rate_to_sleep=MIN_RATE_LIMIT,
+                 max_retries=MAX_RETRIES, default_sleep_time=DEFAULT_SLEEP_TIME):
+        self.base_url = base_url
+        self.rate_limit = None
+        self.rate_limit_reset_ts = None
+        self.session = None
+
+        self.sleep_for_rate = sleep_for_rate
+
+        if min_rate_to_sleep > self.MAX_RATE_LIMIT:
+            msg = "Minimum rate to sleep value exceeded (%d)."
+            msg += "High values might cause the client to sleep forever."
+            msg += "Reset to %d."
+            self.min_rate_to_sleep = self.MAX_RATE_LIMIT
+            logger.warning(msg, min_rate_to_sleep, self.MAX_RATE_LIMIT)
+        else:
+            self.min_rate_to_sleep = min_rate_to_sleep
+
+        self.max_retries = max_retries
+        self.default_sleep_time = default_sleep_time
+
+    def create_http_session(self, headers=None):
+        self.session = requests.Session()
+
+        if headers:
+            self.session.headers.update(headers)
+
+    def close_http_session(self, session):
+        session.close()
+
+    def set_retry_after_status_codes(self, status_codes=None):
```

Comment:
Why is this needed?

Reply:
set_retry_after_status_codes is needed to modify the list of status codes were the retry-after header may appear. By default, those codes are 503, 413, 429, however in same cases (GitHub), the retry-after can appear also with other status codes (403)

Reply:
But, why do we need to do it this way? Why don't you need to modify Retry object and not the instance of this object?

Reply:
You're right, better to modify the instance of the object

Reply:
I tried to modify the instance of the object, but the new error status codes were not taken into account (the test **test_abuse_rate_limit** was failing). So I left the code as it was